### PR TITLE
Add support for generic URI in Storage object

### LINKED ISSUE #826: Support for general web URI address for hosted KFServing model location
/kind feature

**Describe the solution you'd like**

Currently there are 4 types of storage URIs that are accepted in KFServing yaml files:

https://github.com/kubeflow/kfserving/blob/e25758f6da020e3e8ad63f0880ab95de6af8395b/python/kfserving/kfserving/storage.py#L26-L30

(1) GCS
(2) AWS S3
(3) Azure (blob over https://)
(4) local (file://)

I think it would be useful to be able reference a model file hosted anywhere (via general web URI, aka "http://", such as a file on github:

https://github.com/jpatanooga/kubeflow_ops_book_dev/blob/master/kfserving/models/sklearn/iris_sklearn_svm_model.joblib

(in this example the raw file would be downloaded from the url: https://github.com/jpatanooga/kubeflow_ops_book_dev/blob/master/kfserving/models/sklearn/iris_sklearn_svm_model.joblib?raw=true )

This would allow for examples to be executed more efficiently; the KFServing project could even provide a repo of the pre-built models (perhaps), so that example model deployments could be easier to run for new users.

### PR DESCRIPTION
**Pull request by tduffy000**
**What this PR does / why we need it**:
This adds support for a generic uri over HTTP or HTTPS in the Storage [object](https://github.com/kubeflow/kfserving/blob/91165ea3b94f77432029ecc54bec7f4f888f842b/python/kfserving/kfserving/storage.py#L32) responsible for pulling remote model objects, as mentioned in this [issue](https://github.com/kubeflow/kfserving/issues/826). This would allow for the inclusion of serialized models available on the web (e.g. on GitHub, ModelZoo, etc.) in the KFServing framework.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #826

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by tduffy000** [comment]
@googlebot I signed it!

**Comment by tduffy000** [comment]
Seeing this error on the end-2-end test:
```
----------------------------- Captured stderr call -----------------------------
[I 200726 13:45:31 utils:39] Sending request data: {"instances": [[6.8, 2.8, 4.8, 1.4], [6.0, 3.4, 4.5, 1.6]]}
[I 200726 13:45:32 utils:41] Got response code 500, content b'while calling post: Post http://0.0.0.0:8080/v1/models/isvc-logger:predict: dial tcp 0.0.0.0:8080: connect: connection refuse
d\n'
```
In call to the `predict` utility [function](https://github.com/kubeflow/kfserving/blob/91165ea3b94f77432029ecc54bec7f4f888f842b/test/e2e/common/utils.py#L29), is it possibly related to this issue? https://github.com/kubeflow/kfserving/issues/604 (Even though 404 != 500)

Longer:
```   
KFServing.create(isvc)
KFServing.wait_isvc_ready(service_name, namespace=KFSERVING_TEST_NAMESPACE)
>       res = predict(service_name, './data/iris_input.json')
logger/test_logger.py:77:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
common/utils.py:42: in predict
   preds = json.loads(response.content.decode('utf-8'))
/usr/lib/python3.6/json/__init__.py:354: in loads
   return _default_decoder.decode(s)
/usr/lib/python3.6/json/decoder.py:339: in decode
   obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
self = <json.decoder.JSONDecoder object at 0x7ff4648a7320>
s = 'while calling post: Post http://0.0.0.0:8080/v1/models/isvc-logger:predict: dial tcp 0.0.0.0:8080: connect: connection refused\n'
```

File: python/kfserving/kfserving/storage.py [hunk 1]
```diff
@@ -206,6 +211,17 @@ def _download_local(uri, out_dir=None):
             os.symlink(src, dest_path)
         return out_dir
 
+    @staticmethod
+    def _download_from_uri(uri, out_dir=None):
+        fname = uri.split('?')[0].split('/')[-1]
```
Comment (yuzisun) [inline python/kfserving/kfserving/storage.py:216]: you can call `urlparse` and  get the `fname` with `os.path.basename(a.path)`.

**Comment by yuzisun** [comment]
thanks @tduffy000! This is a great feature to add! If you can add an simple example [here](https://github.com/kubeflow/kfserving/tree/master/docs/samples#deploy-inferenceservice-with-cloudpvc-storage) that would be awesome.

**Comment by yuzisun** [comment]
> Seeing this error on the end-2-end test:
> 
> ```
> ----------------------------- Captured stderr call -----------------------------
> [I 200726 13:45:31 utils:39] Sending request data: {"instances": [[6.8, 2.8, 4.8, 1.4], [6.0, 3.4, 4.5, 1.6]]}
> [I 200726 13:45:32 utils:41] Got response code 500, content b'while calling post: Post http://0.0.0.0:8080/v1/models/isvc-logger:predict: dial tcp 0.0.0.0:8080: connect: connection refuse
> d\n'
> ```
> 
> In call to the `predict` utility [function](https://github.com/kubeflow/kfserving/blob/91165ea3b94f77432029ecc54bec7f4f888f842b/test/e2e/common/utils.py#L29), is it possibly related to this issue--> #604 (Even though 404 != 500)
> 
> Longer:
> 
> ```
> KFServing.create(isvc)
> KFServing.wait_isvc_ready(service_name, namespace=KFSERVING_TEST_NAMESPACE)
> >       res = predict(service_name, './data/iris_input.json')
> logger/test_logger.py:77:
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> common/utils.py:42: in predict
>    preds = json.loads(response.content.decode('utf-8'))
> /usr/lib/python3.6/json/__init__.py:354: in loads
>    return _default_decoder.decode(s)
> /usr/lib/python3.6/json/decoder.py:339: in decode
>    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> self = <json.decoder.JSONDecoder object at 0x7ff4648a7320>
> s = 'while calling post: Post http://0.0.0.0:8080/v1/models/isvc-logger:predict: dial tcp 0.0.0.0:8080: connect: connection refused\n'
> ```

I will take a look, I think it is a different issue.

**Comment by yuzisun** [comment]
@tduffy000  the test passed now, the change mostly lgtm left a small comment.

**Comment by ellistarn** [comment]
Apologies. I just rewrote history to fix the contributors list. This PR needs a rebase before it can be reopened.

**Comment by tduffy000** [comment]
@ellistarn I rebased and pushed a change for the comment by @yuzisun can you re-open this? I don't believe I have the permissions to.

**Comment by yuzisun** [comment]
@tduffy000 looks like even PR owner can not reopen the PR, Let's create a new PR then

**Pull request closed without merging**